plugin/pkg/scheduler/factory/factory.go

```diff
@@ -201,11 +203,119 @@ func NewConfigFactory(
+		cache.ResourceEventHandlerFuncs{
+			// MaxPDVolumeCountPredicate: since it relies on the counts of PV.
+			AddFunc:    c.onPvAdd,
+			DeleteFunc: c.onPvDelete,
+		},
+		0,
+	)
+	c.pVLister = pvInformer.Lister()
+
+	// This is for MaxPDVolumeCountPredicate: add/delete PVC will affect counts of PV when it is bound.
+	pvcInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.onPvcAdd,
+			DeleteFunc: c.onPvcDelete,
+		},
+		0,
+	)
+	c.pVCLister = pvcInformer.Lister()
+
+	// This is for ServiceAffinity: it will not be affected by svc add/delete,
+	// except that the selector of the service is updated.
+	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: c.onServiceUpdate,
+		},
+		0,
+	)
+	c.serviceLister = serviceInformer.Lister()
+
+	// NOTE(resouer) Existing equivalence cache should not be affected by add/delete RC/Deployment etc, it only make sense when
+	// pod is scheduled or deleted
 
 	return c
 }
 
+func (c *ConfigFactory) onPvAdd(obj interface{}) {
+	c.invalidatePredicatesForPv(obj.(*v1.PersistentVolume))
+}
+
+func (c *ConfigFactory) onPvDelete(obj interface{}) {
+	var pv *v1.PersistentVolume
+	switch t := obj.(type) {
+	case *v1.PersistentVolume:
+		pv = t
+	case cache.DeletedFinalStateUnknown:
+		var ok bool
+		pv, ok = t.Obj.(*v1.PersistentVolume)
+		if !ok {
+			glog.Errorf("cannot convert to *v1.PersistentVolume: %v", t.Obj)
+			return
+		}
+	default:
+		glog.Errorf("cannot convert to *v1.PersistentVolume: %v", t)
+		return
+	}
+	c.invalidatePredicatesForPv(pv)
+}
+
+func (c *ConfigFactory) invalidatePredicatesForPv(pv *v1.PersistentVolume) {
+	invalidPredicates := sets.NewString("MaxPDVolumeCountPredicate")
+	if pv.Spec.AWSElasticBlockStore != nil {
+		invalidPredicates.Insert("MaxEBSVolumeCount")
+	}
+	if pv.Spec.GCEPersistentDisk != nil {
+		invalidPredicates.Insert("MaxGCEPDVolumeCount")
+	}
+	if pv.Spec.AzureDisk != nil {
+		invalidPredicates.Insert("MaxAzureDiskVolumeCount")
+	}
+	c.equivalencePodCache.InvalidateCachedPredicateItemOfAllNodes(invalidPredicates)
+}
+
+func (c *ConfigFactory) onPvcAdd(obj interface{}) {
+	c.invalidatePredicatesForPvc(obj.(*v1.PersistentVolumeClaim))
+}
+
+func (c *ConfigFactory) onPvcDelete(obj interface{}) {
+	var pvc *v1.PersistentVolumeClaim
+	switch t := obj.(type) {
+	case *v1.PersistentVolumeClaim:
+		pvc = t
+	case cache.DeletedFinalStateUnknown:
+		var ok bool
+		pvc, ok = t.Obj.(*v1.PersistentVolumeClaim)
+		if !ok {
+			glog.Errorf("cannot convert to *v1.PersistentVolumeClaim: %v", t.Obj)
+			return
+		}
+	default:
+		glog.Errorf("cannot convert to *v1.PersistentVolumeClaim: %v", t)
+		return
+	}
+	c.invalidatePredicatesForPvc(pvc)
+}
+
+func (c *ConfigFactory) invalidatePredicatesForPvc(pvc *v1.PersistentVolumeClaim) {
+	if pvc.Spec.VolumeName != "" {
+		c.equivalencePodCache.InvalidateCachedPredicateItemOfAllNodes(sets.NewString("MaxPDVolumeCountPredicate"))
```

Comment:
I'm not very sure how to do this ... since we need to insert value into this set.